crates/ty_project/src/walk.rs

```diff
@@ -258,3 +290,412 @@ pub(crate) enum WalkError {
+                        .next()
+                        .is_none()
+                    {
+                        return PatternMatch::Exclude(ExcludeReason::NoIncludePattern);
+                    }
+                }
+            } else {
+                // If this is a file and there's at least one include pattern but the file doesn't match it,
+                // then the file is excluded. If there are only exclude patterns, than the file should be included.
+                return PatternMatch::Exclude(ExcludeReason::NoIncludePattern);
+            }
+        }
+
+        PatternMatch::None
+    }
+}
+
+impl PartialEq for FilePatterns {
+    fn eq(&self, other: &Self) -> bool {
+        self.patterns == other.patterns
+    }
+}
+
+impl Eq for FilePatterns {}
+
+impl std::fmt::Debug for FilePatterns {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("FilePatterns")
+            .field("patterns", &self.patterns)
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct FilePatternsBuilder {
+    set: GlobSetBuilder,
+    patterns: Vec<FilePattern>,
+    static_prefixes: Option<Vec<SystemPathBuf>>,
+    num_positive: usize,
+}
+
+impl FilePatternsBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            set: GlobSetBuilder::new(),
+            patterns: Vec::new(),
+            static_prefixes: Some(Vec::new()),
+            num_positive: 0,
+        }
+    }
+
+    pub(crate) fn add(&mut self, input: &str) -> Result<&mut Self, globset::Error> {
+        let mut pattern = FilePattern {
+            negated: false,
+            is_only_directory: false,
+            original: input.to_string(),
+        };
+
+        let mut glob = input;
+
+        if let Some(after) = glob.strip_prefix('!') {
+            pattern.negated = true;
+            glob = after;
+        }
+
+        // A pattern ending with a `/` should only match directories. E.g. `src/` only matches directories
+        // whereas `src` matches both files and directories.
+        // We need to remove the `/` to ensure that a path missing the trailing `/` matches.
+        if let Some(before) = glob.strip_suffix('/') {
+            pattern.is_only_directory = true;
+            glob = before;
+
+            // If the slash was escaped, then remove the escape.
+            // See: https://github.com/BurntSushi/ripgrep/issues/2236
+            let trailing_backslashes = glob.chars().rev().filter(|c| *c == '\\').count();
+            if trailing_backslashes % 2 == 1 {
+                glob = &glob[..glob.len() - trailing_backslashes]
+            }
+        }
+
+        // If the last component contains no wildcards or extension, consider it an implicit glob
+        // This turns `src` into `src/**/*`
+        // TODO: Should we also enable this behavior for `is_only_directory` patterns?
+        if is_implicit_glob(glob) && !pattern.negated {
```

Comment:
I think we also need to do this if the pattern's negated or the following won't work:

```
files = ["src", "!src"]
```

There's no issue when we traverse the directory top down, but there's an issue if we match an individual file, e.g `src/a.py`, in which case `src/**` would still match, but `!src` doesn't.